src/v/cluster/tests/cluster_tests.cc

```diff
@@ -170,3 +170,127 @@ FIXTURE_TEST(test_feature_table_snapshots, cluster_test_fixture) {
+    // not. This will be redirected to the controller node, and the deprecated
+    // path in topic_table::apply() should be taken.
+    {
+        auto update = cluster::topic_properties_update{
+          {model::kafka_namespace, topic}};
+        update.properties.get_shadow_indexing().op
+          = cluster::incremental_update_operation::set;
+        update.properties.get_shadow_indexing().value
+          = model::shadow_indexing_mode::full;
+        auto res = node_1->controller->get_topics_frontend()
+                     .local()
+                     .update_topic_properties(
+                       {update}, model::timeout_clock().now() + 5s)
+                     .get();
+
+        BOOST_REQUIRE_EQUAL(res.size(), 1);
+        BOOST_REQUIRE_EQUAL(res[0].ec, cluster::errc::success);
+        auto cfg = node_0->controller->get_topics_state().local().get_topic_cfg(
+          tp_ns);
+        BOOST_REQUIRE(cfg.has_value());
+        const auto& cfg_shadow_indexing
+          = cfg.value().properties.shadow_indexing;
+        BOOST_REQUIRE(cfg_shadow_indexing.has_value());
+        BOOST_REQUIRE_EQUAL(
+          cfg_shadow_indexing.value(), model::shadow_indexing_mode::full);
+    }
+
+    // Simulate a case in which an updated node makes an alter config
+    // call, where shadow_indexing_mode has an update, as well as
+    // remote_read/remote_write. This will be redirected to the controller node,
+    // and the non-deprecated path in topic_table::apply() should be taken.
+    {
+        auto update = cluster::topic_properties_update{
+          {model::kafka_namespace, topic}};
+        update.properties.get_shadow_indexing().op
+          = cluster::incremental_update_operation::set;
+        update.properties.get_shadow_indexing().value
+          = model::shadow_indexing_mode::disabled;
+
+        update.properties.remote_read.op
+          = cluster::incremental_update_operation::set;
+        update.properties.remote_read.value = false;
+
+        update.properties.remote_write.op
+          = cluster::incremental_update_operation::set;
+        update.properties.remote_write.value = false;
+
+        auto res = node_1->controller->get_topics_frontend()
+                     .local()
+                     .update_topic_properties(
+                       {update}, model::timeout_clock().now() + 5s)
+                     .get();
+
+        BOOST_REQUIRE_EQUAL(res.size(), 1);
+        BOOST_REQUIRE_EQUAL(res[0].ec, cluster::errc::success);
+        auto cfg = node_0->controller->get_topics_state().local().get_topic_cfg(
+          tp_ns);
+        BOOST_REQUIRE(cfg.has_value());
+        const auto& cfg_shadow_indexing
+          = cfg.value().properties.shadow_indexing;
+        BOOST_REQUIRE(cfg_shadow_indexing.has_value());
+        BOOST_REQUIRE_EQUAL(
+          cfg_shadow_indexing.value(), model::shadow_indexing_mode::disabled);
+    }
+
+    // Simulate a case in which an updated node makes an alter config
```

Comment:
The two options I can think of are:
* Using kcl to send requests similar to this test: https://github.com/redpanda-data/redpanda/blob/4c2c24c468048905a00d4f52bc546e97ad88bcbf/tests/rptest/tests/topic_creation_test.py#L626-L649 This works for both `AlterConfigs` and `IncrementalAlterConfigs` but you would need to make a small change to our kcl.py wrapper to make the target host configurable here https://github.com/redpanda-data/redpanda/blob/4c2c24c468048905a00d4f52bc546e97ad88bcbf/tests/rptest/clients/kcl.py#L383-L385
* Using rpk. Similarly, you would need to pipe through a target host down from https://github.com/redpanda-data/redpanda/blob/4c2c24c468048905a00d4f52bc546e97ad88bcbf/tests/rptest/clients/rpk.py#L750 to https://github.com/redpanda-data/redpanda/blob/4c2c24c468048905a00d4f52bc546e97ad88bcbf/tests/rptest/clients/rpk.py#L1286

Also, just to clarify, when I suggested using an end-to-end test I just meant using an external client to change the configs, not necessarily an EndToEndTest.

Reply:
Added `test_alter_config_shadow_indexing_mixed_node` to `[alter_topic_configuration_test.py](https://github.com/redpanda-data/redpanda/compare/21ce4251ea1688ab3243e4c82a70908153f90635..ae4d02c978f6ae8d35075d23a721474db44a6b8e#diff-0879970555c08129c60c644a81e0f1a62799e61d8e9f026089502075e5db4235)`

Reply:
For some reason `rpk` (or maybe `franz-go`) isn't respecting specified brokers in `_kafka_conn_settings()`, with the log line `01:05:25.003  DEBUG  sharded request  {"req": "DescribeConfigs", "destinations": ["any"]}`.

Using `KCL` works as expected.

Reply:
cc: @twmb in case you have any thoughts about allowing `rpk` commands to specify brokers in the same way as `kafka` commands with the parameter `seed_brokers=...`